lib/jsdom/living/per-document-resource-loader.js

```diff
@@ -0,0 +1,119 @@
+"use strict";
+
+const wrapCookieJarForRequest = require("./helpers/wrap-cookie-jar-for-request");
+
+class PerDocumentResourceLoader {
+  constructor(document) {
+    this._document = document;
+    this._defaultEncoding = document._encoding;
+    this._resourceLoader = document._defaultView ? document._defaultView._resourceLoader : null;
+    this._requestManager = document._requestManager;
+    this._queue = document._queue;
+    this._deferQueue = document._deferQueue;
+    this._asyncQueue = document._asyncQueue;
+
+    // TODO: Remove when _hasFeature is no longer necessary
+    this._implementation = document._implementation;
+  }
+
+  fetch(url, options) {
+    const { element } = options;
+
+    if (!this._implementation._hasFeature("FetchExternalResources", element.tagName.toLowerCase())) {
+      return null;
+    }
+
+    if (this._implementation._hasFeature("SkipExternalResources", url)) {
+      return null;
+    }
+
+    const requestOptions = {
+      jar: wrapCookieJarForRequest(this._document._cookieJar),
+      proxy: this._document._proxy,
+      strictSSL: typeof this._document._strictSSL === "boolean" ? this._document._strictSSL : true
+    };
+
+    options = Object.assign({}, options, { requestOptions, referrer: this._document.URL });
+
+    const request = this._resourceLoader.fetch(
+      url,
+      options
+    );
+
+    this._requestManager.add(request);
+
+    const ev = this._document.createEvent("HTMLEvents");
+
+    const onError = error => {
+      this._requestManager.remove(request);
+
+      if (options.onError) {
+        options.onError(error);
+      }
+
+      // TODO: Create events in the modern way https://github.com/jsdom/jsdom/pull/2279#discussion_r199969734
+      ev.initEvent("error", false, false);
+      element.dispatchEvent(ev);
+
+      const err = new Error(`Could not load ${element.localName}: "${url}"`);
+      err.detail = error;
+      err.type = "resource loading";
+
+      this._document._defaultView._virtualConsole.emit("jsdomError", err);
+
+      return Promise.resolve();
+    };
+
+    const onLoad = data => {
+      this._requestManager.remove(request);
+
+      this._addCookies(url, request.response ? request.response.headers : {});
+
+      try {
+        const result = options.onLoad ? options.onLoad(data) : undefined;
+
+        return Promise.resolve(result)
+          .then(() => {
+            // TODO: Create events in the modern way https://github.com/jsdom/jsdom/pull/2279#discussion_r199969734
+            ev.initEvent("load", false, false);
+            element.dispatchEvent(ev);
+
+            return Promise.resolve();
+          })
+          .catch(err => {
+            return onError(err);
+          });
+      } catch (err) {
+        return onError(err);
+      }
+    };
+
+    if (element.localName === "script" && element.getAttribute("async") !== null) {
```

Comment:
Adding support for `async` scripts is great, but `HTMLScriptElement` also needs to be updated to handle the attribute. You can do this by replacing the `async` attribute line [here](https://github.com/jsdom/jsdom/blob/master/lib/jsdom/living/nodes/HTMLScriptElement.webidl#L7) with:

`[CEReactions, Reflect] attribute boolean async;`

...and run `yarn convert-idl` to process the WebIDL file.

We'll also need test coverage since this is a new feature. Some of the `async` tests in https://github.com/jsdom/jsdom/blob/master/test/web-platform-tests/to-run.yaml#L589 should be passing with these changes, so you can enable them by removing them from the list.